### PR TITLE
Forced CRD sharing from System tenant

### PR DESCRIPTION
**Pull request by chenqianfzh**
**What type of PR is this?**
> /kind feature

### What is in this PR?
The feature of forced CRD sharig. When the system tenant creates a CRD with annotation "futurewei.k8s.io/crd-sharing-policy": "forced". This CRD will be usable by every tenant. This sharing is forced, which means that the regular tenants CANNOT create another CRD to override the system version.

### What is NOT in this PR?
As the details of the forced CRD sharing feature is still on discussion and subject to change, no tests are included in this PR.

### Tests Done in my Devbox
I started the cluster with Arktos-up.sh. Two tenants, qian and peng, are created. 

#### regular tenants can see the system-forced-sharing CRD just like their own CRDs:
```
prompt$ kubectl create -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/forced_crd.yaml
customresourcedefinition.apiextensions.k8s.io/burgers.futurewei.k8s.io created
  
prompt$ kubectl apply -f ~/gopath/qianc/arktos/staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml  --tenant qian
customresourcedefinition.apiextensions.k8s.io/foos.samplecontroller.k8s.io created
  
prompt$ kubectl get crds --tenant qian
NAME                           CREATED AT
burgers.futurewei.k8s.io       2020-06-01T04:29:55Z
foos.samplecontroller.k8s.io   2020-06-01T04:29:56Z
  
prompt$ kubectl get crds --tenant peng
NAME                       CREATED AT
burgers.futurewei.k8s.io   2020-06-01T04:29:55Z
```
### A regular tenant's trial to create a CRD with the same name of a system-forced-sharing CRD will be blocked.
```
prompt$ kubectl create -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/forced_crd.yaml --tenant qian
Error from server: error when creating "/home/qianchen/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/forced_crd.yaml": There is already a system forced CRD with the name burgers.futurewei.k8s.io 
```

```
qianchen@qianchen-VirtualBox:~/gopath/qianc/arktos$ kubectl apply -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/forced_crd.yaml  --tenant peng
Error from server: error when applying patch:
...
for: "/home/qianchen/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/forced_crd.yaml": burgers.arktos.futurewei.com is a system CRD, you cannot overwrite it.
```

### Tenants can create custom resources based on the CRD. 
```
prompt$ kubectl apply -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/turkey-burger.yaml --tenant qian
burger.futurewei.k8s.io/turkey-burger created
  
prompt$ kubectl apply -f  ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/beef-burger.yaml  --tenant peng
burger.futurewei.k8s.io/beef-burger created
```

### You can list the custom resources across all the tenants, or based on a specific tenant, just like how you work with a native resource
```
prompt$ kubectl get burgers --all-tenants --all-namespaces
TENANT   NAMESPACE   NAME            AGE
peng     default     beef-burger     1s
qian     default     turkey-burger   1s
  
prompt$ kubectl get burgers --tenant peng
NAME          AGE
beef-burger   1s
  
prompt$ kubectl get burgers --tenant qian
NAME            AGE
turkey-burger   7s
```

### When you delete a system-forced-sharing CRD, all the custom resouces of this type will be gone along with the CRD itself
```
prompt$ kubectl delete -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/forced_crd.yaml
customresourcedefinition.apiextensions.k8s.io "burgers.futurewei.k8s.io" deleted

prompt$ kubectl get burgers --all-tenants --all-namespaces
Error from server (NotFound): Unable to list "futurewei.k8s.io/v1alpha1, Resource=burgers": the server could not find the requested resource (get burgers.futurewei.k8s.io)
  
prompt$ kubectl get burgers --tenant peng
Error from server (NotFound): Unable to list "futurewei.k8s.io/v1alpha1, Resource=burgers": the server could not find the requested resource (get burgers.futurewei.k8s.io)
  
prompt$ kubectl get burgers --tenant qian
Error from server (NotFound): Unable to list "futurewei.k8s.io/v1alpha1, Resource=burgers": the server could not find the requested resource (get burgers.futurewei.k8s.io)
```

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go [hunk 1]
```diff
@@ -19,24 +20,33 @@ package customresourcedefinition
...
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
 )
 
 // rest implements a RESTStorage for API services against etcd
 type REST struct {
 	*genericregistry.Store
 }
 
+const (
+	crdSharingPolicyAnnotation = "futurewei.k8s.io/crd-sharing-policy"
```
Comment (h-w-chen) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go:46]: not major issue - considering arktos.futurewei.com already is used somewhere else,  "arktos.futurewei.com/crd-sharing-policy" seems consistent.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go [hunk 2]
```diff
@@ -66,6 +76,102 @@ func (r *REST) ShortNames() []string {
...
+	}
+
+	resultList := &apiextensions.CustomResourceDefinitionList{
+		TypeMeta: tenantCrdList.TypeMeta,
+		ListMeta: tenantCrdList.ListMeta,
+		Items:    append(systemCrds, tenantCrds...)}
+	return resultList, nil
+}
+
+func (r *REST) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
```
[the remark this answers is not in the document]
Comment (chenqianfzh) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go:154]: Yes, it is the behavior given in the PR.

See the PR description:

### A regular tenant's trial to create a CRD with the same name of a system-forced-sharing CRD will be blocked.
```
prompt$ kubectl create -f ~/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/forced_crd.yaml --tenant qian
Error from server: error when creating "/home/qianchen/gopath/qianc/arktos/test/yaml/multi_tenancy_examples/crd/forced_crd.yaml": There is already a system forced CRD with the name burgers.futurewei.k8s.io

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go [hunk 3]
```diff
@@ -66,6 +76,102 @@ func (r *REST) ShortNames() []string {
 	return []string{"crd", "crds"}
 }
 
+// try to retrieve the forced version of CRD under the system teant first.
```
[the remark this answers is not in the document]
Comment (chenqianfzh) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go:79]: updated

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go (hunk 1 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go:46]: updated.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go [hunk 4]
```diff
@@ -66,6 +76,102 @@ func (r *REST) ShortNames() []string {
...
+		return nil, err
+	}
+
+	tenantCrdList, ok := tenantList.(*apiextensions.CustomResourceDefinitionList)
+	if !ok {
+		return nil, fmt.Errorf("Failed to convert the object to CRD list")
+	}
+
+	tenantCrds := []apiextensions.CustomResourceDefinition{}
+	for _, crd := range tenantCrdList.Items {
```
[the remark this answers is not in the document]
Comment (chenqianfzh) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go:135]: Changed.

**Pull request merged**